apollo-router/src/plugins/rhai/engine.rs

```diff
@@ -0,0 +1,1201 @@
+use std::path::PathBuf;
+use std::str::FromStr;
+use std::sync::Arc;
+use std::sync::Mutex;
+use std::time::SystemTime;
+
+use http::header::InvalidHeaderName;
+use http::uri::Authority;
+use http::uri::Parts;
+use http::uri::PathAndQuery;
+use http::Uri;
+use rhai::module_resolvers::FileModuleResolver;
+use rhai::plugin::*;
+use rhai::serde::from_dynamic;
+use rhai::serde::to_dynamic;
+use rhai::Array;
+use rhai::Dynamic;
+use rhai::Engine;
+use rhai::EvalAltResult;
+use rhai::FnPtr;
+use rhai::Instant;
+use rhai::Map;
+use rhai::Scope;
+use rhai::AST;
+use tower::BoxError;
+use uuid::Uuid;
+
+use super::execution;
+use super::subgraph;
+use super::supergraph;
+use super::Rhai;
+use super::ServiceStep;
+use crate::graphql::Request;
+use crate::graphql::Response;
+use crate::http_ext;
+use crate::plugins::authentication::APOLLO_AUTHENTICATION_JWT_CLAIMS;
+use crate::Context;
+
+const CANNOT_ACCESS_HEADERS_ON_A_DEFERRED_RESPONSE: &str =
+    "cannot access headers on a deferred response";
+
+pub(super) trait OptionDance<T> {
+    fn with_mut<R>(&self, f: impl FnOnce(&mut T) -> R) -> R;
+
+    fn replace(&self, f: impl FnOnce(T) -> T);
+
+    fn take_unwrap(self) -> T;
+}
+
+pub(super) type SharedMut<T> = rhai::Shared<Mutex<Option<T>>>;
+
+impl<T> OptionDance<T> for SharedMut<T> {
+    fn with_mut<R>(&self, f: impl FnOnce(&mut T) -> R) -> R {
+        let mut guard = self.lock().expect("poisoned mutex");
+        f(guard.as_mut().expect("re-entrant option dance"))
+    }
+
+    fn replace(&self, f: impl FnOnce(T) -> T) {
+        let mut guard = self.lock().expect("poisoned mutex");
+        *guard = Some(f(guard.take().expect("re-entrant option dance")))
+    }
+
+    fn take_unwrap(self) -> T {
+        match Arc::try_unwrap(self) {
+            Ok(mutex) => mutex.into_inner().expect("poisoned mutex"),
+
+            // TODO: Should we assume the Arc refcount is 1
+            // and use `try_unwrap().expect("shared ownership")` instead of this fallback ?
+            Err(arc) => arc.lock().expect("poisoned mutex").take(),
+        }
+        .expect("re-entrant option dance")
+    }
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct RhaiHeaderMap(pub(crate) http::HeaderMap);
```

Comment:
I'm not clear why we need to introduce this new type? Could you explain why you've added it?

Reply:
I added it because `HeaderMap` is registered as an iterator that produces `(Option<HeaderName>, HeaderValue)`, and I could not use that as a type exposed by a module. So I created the `HeaderPair` struct with the same content, and made the `RhaiHeaderMap` wrapper that produces `HeaderPair` when iterated on

Reply:
if you see a better way to do that, I'm happy to make the change, because I agree that it feels weird